src/date.c

```diff
@@ -1,18 +1,1006 @@
+        d_argtoint(v7, &arg, &ret_time);
+      }
+    } else {
+      /* 2+ paramaters - should be parts of a date */
+      struct dtimepartsarr a;
+      int i;
+      
+      memset(&a, 0, sizeof(a));
+      
+      for(i=0; i < cargs; i++) {
+        val_t val = v7_array_at(v7, args, i);
+        if(!d_argtoint(v7, &val, &a.args[i])) {
+          break;
+        }
+      }
+      
+      if(i>=cargs) {
+        if(a.args[tpdate] == 0 ) {
+          a.args[tpdate] = 1; /* If date is supplied then let 
+                               dt be ToNumber(date); else let dt be 1. */
+        }
+      
+        if(a.args[tpyear] >=0 && a.args[tpyear] <=99) {
+          a.args[tpyear] += 1900;
+          /* If y is not NaN and 0 <= ToInteger(y) <= 99,
+                  then let yr be 1900+ToInteger(y); otherwise, let yr be y. */
+        }
+
+        ret_time = ecma_UTC(d_changepartoftime(0, &a, 0, d_gmktime));
+      }
+    }
+  
+    v7_to_object(this_obj)->prototype = v7_to_object(v7->date_prototype);
+    v7_set_property(v7, this_obj, "", 0, V7_PROPERTY_HIDDEN,
+                                        v7_create_number(ret_time));
+    return this_obj;
+  } else {
+    /* according to 15.9.2.1 we should ignore all 
+                          parameters in case of function-call */
+    struct timeparts tp;
+    char buf[50];
+    int len;
+    
+    d_gettime(&ret_time);
+    d_localtime(&ret_time, &tp);
+    len = d_tptostr(&tp, buf, 1);
+    
+    return v7_create_string(v7, buf, len, 1);
+  }
+}
+
+/*++++ toXXXString functions ++++*/
+static val_t Date_toISOString(struct v7 *v7, val_t this_obj, val_t args) {
+  char buf[30];
+  etime_t time; int len;
+  (void)args;
+  
+  d_isobjvalidforstirng(v7, this_obj);
+  
+  time = v7_to_double(i_value_of(v7, this_obj));
+  len = d_timetoISOstr(&time, buf, sizeof(buf));
+  
+  return v7_create_string(v7, buf, len, 1);
+}
+
+/* using macros to avoid copy-paste technic */
+#define DEF_TOSTR(funcname, breaktimefunc,tostrfunc, addtz) \
+  static val_t Date_to##funcname(struct v7 *v7, val_t this_obj, val_t args) {  \
+    (void)args; \
+    return d_tostring(v7, this_obj, breaktimefunc, tostrfunc, addtz); \
+  }
+
+DEF_TOSTR(UTCString, d_gmtime, d_tptostr, 0)
+DEF_TOSTR(String, d_localtime, d_tptostr, 1)
+DEF_TOSTR(DateString, d_localtime, d_tptodatestr, 1)
+DEF_TOSTR(TimeString, d_localtime, d_tptotimestr, 1)
+
+
+#define DEF_TOLOCALESTR(funcname, frm) \
+  static val_t Date_to##funcname(struct v7 *v7, val_t this_obj, val_t args) { \
+    (void)args; \
+    return d_tolocalestr(v7, this_obj, frm);  \
+  }
+
+DEF_TOLOCALESTR(LocaleString, "%c")
+DEF_TOLOCALESTR(LocaleDateString, "%x")
+DEF_TOLOCALESTR(LocaleTimeString, "%X")
+
+/*++++ get functions ++++*/
+
+#define DEF_GET_TP_FUNC(funcName, tpmember, breaktimefunc) \
+  static val_t Date_get##funcName(struct v7 *v7, val_t this_obj, val_t args) { \
+    struct timeparts tp; \
+    return v7_create_number(d_isnumberNAN(v7, this_obj)? NAN: d_getTP(v7, \
```

Comment:
why not invoke `i_value_of` only once here?

Rationale:

`i_value_of` takes a value and returns itself if it's a scalar, the result of the method `valueOf()` if it's an object (e.g. a custom method or the builtins such as Number.valueOf() which extracts a scalar from the object wrappers).

In the case of `Date`, calling `i_value_of` on `this_obj` in each method is just a way to extract it's hidden property, like we do for `Number`, `Boolean` and friends.

There is no need to push down this detail into helpers such as `d_isnumberNAN`, especially since by not doing so we'll also remove unnecessary method invocations.